### PR TITLE
separated scaler & invere scale

### PR DESCRIPTION
**Pull request by jinkwon711**
I added inverse scaling.

### CODE REVIEW

File: Keras/klab-12-4-rnn_deep_prediction.py [hunk 1]
```diff
@@ -21,11 +22,10 @@
 xy = xy[::-1]  # reverse order (chronically ordered)
 
 # very important. It does not work without it.
-scaler = MinMaxScaler(feature_range=(0, 1))
-xy = scaler.fit_transform(xy)
-
-x = xy
-y = xy[:, [-1]]  # Close as label
+X_scaler = MinMaxScaler(feature_range=(0, 1))
+Y_scaler = MinMaxScaler(feature_range=(0, 1))
```
Comment (kkweon) [inline Keras/klab-12-4-rnn_deep_prediction.py]: Can you undo MinMaxScaler on `Y`?

File: Keras/klab-12-4-rnn_deep_prediction.py [hunk 2]
```diff
@@ -21,11 +22,10 @@
...
+x = X_scaler.fit_transform(xy)
+y = Y_scaler.fit_transform(xy[:, [-1]])  # Close as label
```
Comment (kkweon) [inline Keras/klab-12-4-rnn_deep_prediction.py]: Same.

File: Keras/klab-12-4-rnn_deep_prediction.py [hunk 3]
```diff
@@ -60,16 +60,16 @@
...
-model.fit(trainX, trainY, epochs=200)
+model.fit(trainX, trainY, epochs=20)
 
 # make predictions
 testPredict = model.predict(testX)
 
 # inverse values
-# testPredict = scaler.transform(testPredict)
-# testY = scaler.transform(testY)
+testPredict = Y_scaler.inverse_transform(testPredict)
```
Comment (kkweon) [inline Keras/klab-12-4-rnn_deep_prediction.py]: no scaling on y

File: Keras/klab-12-4-rnn_deep_prediction.py (hunk 1 above)
Comment (jinkwon711) [inline Keras/klab-12-4-rnn_deep_prediction.py]: Thanks for the comment. 
Is it okay to not do the scale process on Y?
In the original state, scaler was used scaler for xy, which include both X_data and Y_data
So I thought that it was required. 
if we do not scale Y, isn't the Y value too large for the X_data?

Comment (kkweon) [inline Keras/klab-12-4-rnn_deep_prediction.py]: Yes.
In fact, we should never touch `y` unless there is a very specific reason such as `y` being too large for a float32 variable. Otherwise, it's misleading because its error is not in a true scale and we would never know how big our error is.

Reference
- https://stats.stackexchange.com/questions/111467/is-it-necessary-to-scale-the-target-value-in-addition-to-scaling-features-for-re

**Comment by kkweon** [comment]
I carefully looked over the [original file](https://github.com/hunkim/DeepLearningZeroToAll/blob/master/lab-12-5-rnn_stock_prediction.py)

It has a major flaw that it normalizes the whole dataset even before doing the train/test split.
That's why it works well because it's non-stationary data and it uses the future information to predict a current price.

It will require lots of modifications such as making it stationary.
Wow, I'm surprised I just found out this issue.
I'm going to review this issue soon again.

**Comment by hunkim** [comment]
Thanks for your help.

**Comment by hunkim** [comment]
"normalizes the whole dataset" to simplify the process, but I don't think
it's not too different if you do normalize train and test separately using
training normalization information. And this is not really using future
data. Can you try separate normalization and check the results?

"That's why it works well " I am not sure current model works well. This is
more/less the same as simple linear regression.

**Comment by kkweon** [comment]
Yes I will take a look soon.

I believe the problem is that data is non stationary. Its mean and variance is constantly changing. Since it is time series data, we can't make a regression without making it stationary. 

If it is not stationary, it is not possible to predict because the model is a complex version of moving average (moving average only shows the history not the future)

But does min max scaling make data stationary?
I am not sure.


Similar reference:
"Running the data on a single point-by-point prediction as mentioned above gives something that matches the returns pretty closely. But this is deceptive! Why? Well if you look more closely, the prediction line is made up of singular prediction points that have had the whole prior true history window behind them. Because of that, the network doesn’t need to know much about the time series itself other than that each next point most likely won’t be too far from the last point. So even if it gets the prediction for the point wrong, the next prediction will then factor in the true history and disregard the incorrect prediction, yet again allowing for an error to be made."
http://www.jakob-aungiers.com/articles/a/LSTM-Neural-Network-for-Time-Series-Prediction

**Comment by kkweon** [comment]
I will close this since the code needs refactoring.
